mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -117,29 +128,30 @@ struct ThreadwiseGemmRewritePattern
     {
       OpBuilder::InsertionGuard guard(b);
       b.setInsertionPointToStart(gemmLoop.getBody());
-      Value aVal = b.create<InBoundsLoadOp>(
-          loc, dataType, bufferA, gemmLoop.getLowerCoords(/*domain=*/0));
-      Value bVal = b.create<InBoundsLoadOp>(
-          loc, dataType, bufferB, gemmLoop.getLowerCoords(/*domain=*/1));
-      Value mul;
-      if (dataType.isa<IntegerType>())
-        mul = b.create<MulIOp>(loc, aVal, bVal);
-      else if (dataType.isa<FloatType>())
-        mul = b.create<MulFOp>(loc, aVal, bVal);
-      else
+      // These are vector::TransferRead ops so they always return a vector
+      // result so that FMA doesn't complain
+      Value aVal = b.create<vector::TransferReadOp>(
+          loc, abType, bufferA, gemmLoop.getLowerCoords(/*domain=*/0),
+          /*inBounds=*/ArrayRef<bool>(true));
+      Value bVal = b.create<vector::TransferReadOp>(
+          loc, abType, bufferB, gemmLoop.getLowerCoords(/*domain=*/1),
+          /*inBounds=*/ArrayRef<bool>(true));
+      ValueRange cCoords = gemmLoop.getLowerCoords(/*domain=*/2);
+      Value cVal =
+          b.create<vector::TransferReadOp>(loc, cType, bufferC, cCoords,
+                                           /*inBounds=*/ArrayRef<bool>(true));
+      Value result;
+      if (dataType.isa<IntegerType>()) {
+        Value mul = b.create<MulIOp>(loc, aVal, bVal);
+        result = b.create<AddIOp>(loc, mul, cVal);
+      } else if (dataType.isa<FloatType>()) {
+        result = b.create<vector::FMAOp>(loc, aVal, bVal, cVal);
```

Comment:
This is a different approach than xdlops end of "force" it to use a specific intrinsic. In terms of hit or miss rate, how much would this actually ended up in the dot instructions? I think this is a reasonable patch so that nonxdlops kernel can "occasionally" use dot ops when size happen to matches, but ultimately we'd need a new threadwisegemm to do it exclusively using that.

Reply:
Ok, so, this doesn't do dot instructions at all. Code-structure wise, I'm pretty sure blockwise-to-threadwise extends everything to F32/I32 before we reach this point. More fundamentally, I'm not sure the compiler will optimize 
```
%v1 = fma %a0, %b0 : f16
%v2 = fma %a1, %b1 : f16
%v1_ext = ext %v1 : f32
%v2_ext = ext %v2 : f32
%acc= fadd %acc, %v1_ext, %v2_ext : f32
```
to a dot product